src/engine/platform/es5506.cpp

```diff
@@ -0,0 +1,845 @@
+      }
+      chan[c.chan].inPorta=c.value;
+      break;
+    case DIV_CMD_SAMPLE_POS: {
+      if (chan[c.chan].useWave) break;
+      if (chan[c.chan].active) {
+        unsigned int pos=chan[c.chan].pcm.reversed?(chan[c.chan].pcm.length-c.value):c.value;
+        if ((chan[c.chan].pcm.reversed && pos>0) || ((!chan[c.chan].pcm.reversed) && pos<chan[c.chan].pcm.length)) {
+          pageWrite(0x20|c.chan,0x03,chan[c.chan].pcm.start+(pos<<11));
+        }
+      }
+      break;
+    }
+    case DIV_CMD_GET_VOLMAX:
+      return 255;
+      break;
+    case DIV_ALWAYS_SET_VOLUME:
+      return 1;
+      break;
+    default:
+      break;
+  }
+  return 1;
+}
+
+void DivPlatformES5506::muteChannel(int ch, bool mute) {
+  isMuted[ch]=mute;
+  es5506.set_mute(ch,mute);
+}
+
+void DivPlatformES5506::forceIns() {
+  for (int i=0; i<=chanMax; i++) {
+    chan[i].insChanged=true;
+    chan[i].freqChanged=true;
+    chan[i].volChanged=true;
+    chan[i].filterChanged.changed=(unsigned char)(~0);
+    chan[i].envChanged.changed=(unsigned char)(~0);
+    chan[i].sample=-1;
+  }
+}
+
+void* DivPlatformES5506::getChanState(int ch) {
+  return &chan[ch];
+}
+
+void DivPlatformES5506::reset() {
+  while (!hostIntf32.empty()) hostIntf32.pop();
+  while (!hostIntf8.empty()) hostIntf8.pop();
+  for (int i=0; i<32; i++) {
+    chan[i]=DivPlatformES5506::Channel();
+    chan[i].std.setEngine(parent);
+  }
+  es5506.reset();
+  for (int i=0; i<32; i++) {
+    es5506.set_mute(i,isMuted[i]);
+  }
+
+  cycle=0;
+  curPage=0;
+  maskedVal=0;
+  irqv=0x80;
+  isMasked=false;
+  isReaded=false;
+  irqTrigger=false;
+  chanMax=initChanMax;
+
+  pageWriteMask(0x00,0x60,0x0b,chanMax);
+  pageWriteMask(0x00,0x60,0x0b,0x1f);
+  pageWriteMask(0x20,0x60,0x0a,0x01);
+  pageWriteMask(0x20,0x60,0x0b,0x11);
+  pageWriteMask(0x20,0x60,0x0c,0x20);
+  pageWriteMask(0x00,0x60,0x0c,0x08); // Reset serial output
+}
+
+bool DivPlatformES5506::isStereo() {
+  return true;
+}
+
+bool DivPlatformES5506::keyOffAffectsArp(int ch) {
+  return true;
+}
+
+void DivPlatformES5506::notifyInsChange(int ins) {
+  for (int i=0; i<32; i++) {
+    if (chan[i].ins==ins) {
+      chan[i].insChanged=true;
+    }
+  }
+}
+
+void DivPlatformES5506::notifyWaveChange(int wave) {
+  // TODO when wavetables are added
+  // TODO they probably won't be added unless the samples reside in RAM
+}
+
+void DivPlatformES5506::notifyInsDeletion(void* ins) {
+  for (int i=0; i<32; i++) {
+    chan[i].std.notifyInsDeletion((DivInstrument*)ins);
+  }
+}
+
+void DivPlatformES5506::setFlags(unsigned int flags) {
+  initChanMax=MAX(4,flags&0x1f);
+  chanMax=initChanMax;
+  pageWriteMask(0x00,0x60,0x0b,chanMax);
+}
+
+void DivPlatformES5506::poke(unsigned int addr, unsigned short val) {
+  immWrite(addr, val);
+}
+
+void DivPlatformES5506::poke(std::vector<DivRegWrite>& wlist) {
+  for (DivRegWrite& i: wlist) immWrite(i.addr,i.val);
+}
+
+unsigned char* DivPlatformES5506::getRegisterPool() {
```

Comment:
This function may end up being slow. Are you sure there isn't a way to directly feed the register pool?

Reply:
but chip itself overwrites some registers at runtime when each channel is playing. (ex: filter storage register, phase accumulator, stop bit (keyon bit))